domain/model/mesh_results.py

```diff
@@ -113,29 +117,36 @@ class ConnectionMatrix:
     """
 
     def __init__(self, rows: List[MeshRow]) -> None:
-        agent_ids: List[AgentID] = []
         connections: Dict[AgentID, Dict[AgentID, MeshColumn]] = {}
         for row in rows:
-            agent_ids.append(row.agent_id)
             connections[row.agent_id] = {}
             for col in row.columns:
                 connections[row.agent_id][col.agent_id] = col
         self._connections = connections
-        self.agent_ids = sorted(agent_ids)
         self.connection_timestamp_lowest, self.connection_timestamp_highest = self._get_lowest_highest_timestamp()
 
+    def num_connections_with_health_data(self) -> int:
+        count = 0
+        for row in self._connections.values():
+            for conn in row.values():
+                if conn.has_data():
+                    count += 1
+        return count
+
     def incremental_update(self, src: ConnectionMatrix) -> None:
         """
         Update with src connections, add new connections if any, don't remove anything.
         Prerequisite: agents configuration hasn't change.
         """
 
         for from_agent_id in src._connections.keys():
-            dst_row = self._connections[from_agent_id]  # get or insert
+            dst_row = self._connections.get(from_agent_id, {})  # get or create dst_row
```

Comment:
The previous approach would work if `self._connections` was a [defaultdict](https://docs.python.org/3/library/collections.html#collections.defaultdict).